AsyncDisplayKit/ASCollectionView.mm

```diff
@@ -406,7 +397,47 @@ - (void)setAsyncDelegate:(id<ASCollectionViewDelegate>)asyncDelegate
 
   super.delegate = (id<UICollectionViewDelegate>)_proxyDelegate;
   
-  [_layoutInspector didChangeCollectionViewDelegate:asyncDelegate];
+  if (_layoutInspectorFlags.layoutInspectorDidChangeCollectionViewDelegate) {
+    [_layoutInspector didChangeCollectionViewDelegate:asyncDelegate];
+  }
+}
+
+- (void)setCollectionViewLayout:(UICollectionViewLayout *)collectionViewLayout
+{
+  [super setCollectionViewLayout:collectionViewLayout];
+  
+  // Trigger recreation of layout inspector with new collection view layout
+  if (_layoutInspector != nil) {
+    _layoutInspector = nil;
+    [self layoutInspector];
+  }
+}
+
+- (void)setLayoutInspector:(id<ASCollectionViewLayoutInspecting>)layoutInspector
+{
+  _layoutInspector = layoutInspector;
+  
+  _layoutInspectorFlags.layoutInspectorDidChangeCollectionViewDataSource = [_layoutInspector respondsToSelector:@selector(didChangeCollectionViewDataSource:)];
+  _layoutInspectorFlags.layoutInspectorDidChangeCollectionViewDelegate = [_layoutInspector respondsToSelector:@selector(didChangeCollectionViewDelegate:)];
```

Comment:
@levi I don't know if we should cache the calls in here as I said in the other comment we need to be careful if the `_layoutInspector` changes to call the setter